Update the routing data structure

kellymclaughlin opened the pull request:
Update the routing to use a trie as the underlying data structure rather
than the lists. The goal of these changes are to improve both the space
efficiency of the routing information and the efficiency when performing
lookups to match the resource for a route.

I have created a criterion benchmark suite that I used to test the 
performance of these changes and it can be found [here](https://github.com/kellymclaughlin/airship-route-benchmark). The results I have gotten so far indicate this
change will have a nice benefit to routing performance, but I would welcome 
anyone to run some tests as well to verify my conclusion. Here are some links to the 
results I obtained:

* Results with current `master`: https://dl.dropboxusercontent.com/u/14268813/airship-route-bench-old.html
* Results using this branch: https://dl.dropboxusercontent.com/u/14268813/airship-route-bench-new.html

I have used this branch to run the integration tests of some of our internal 
Airship projects to verify the functionality of the changes and so far I have 
found it to work as expected and be fully compatible with existing Airship 
apps, but would certainly be happy for others to try as well.

kellymclaughlin committed:
Update the routing data structure

Update the routing to use a trie as the underlying data structure rather
than the lists. The goal of these changes are to improve both the space
efficiency of the routing information and the efficiency when performing
lookups to match the resource for a route.

lseppala commented on src/Airship/Resource.hs:
```diff
@@ -151,3 +152,8 @@ defaultResource = Resource { allowMissingPost          = return False
                            }
+
+
+instance Monad m => Monoid (Resource m) where
```
This instance violates the following Monoid laws:

```
mappend mempty x = x
mconcat = foldr mappend mempty
```

 We could instead use the Monoid instance of `First` and the function `fromMaybe defaultResource . getFirst` to achieve the same.

lseppala commented on src/Airship/Internal/Route.hs:
```diff
@@ -32,7 +42,25 @@ data BoundOrUnbound = Bound Text
+                 | RouteMatchOrVar (Resource m) [Text]
+                 | Wildcard (Resource m)
+
+instance Monad m => Monoid (RouteLeaf m) where
```
This violates _all_ of the Monoid laws. Could we instead define a function to achieve the same matching?

lseppala reviewed (changes requested):
I think it's important to maintain the guarantees given by a lawful type class.

charleso commented:
Agree with the monoid comments, any reason for the instances in the first place?

Any chance we could add some tests (or better yet properties) for the routing? It feels like it's fairly testable. Let me know if you need a hand, I'd be happy to help.

kellymclaughlin committed:
Rework so bad Monoid instances are unnecessary

kellymclaughlin commented:
I've reworked the changes to remove the offending `Monoid` instances. My benchmarking shows the performance characteristics to be nearly identical after this change compared to before.

I think the idea of a property to test the routing is a good one. I'll work on that as I have time.

kellymclaughlin committed:
Fix a few regressions and bugs

* Fix a regression introduced in the previous commit where the
  conversion of the routes list to a trie resolved key conflicts by
  having the early value shadow any later values. Add custom conflict
  resolution handling to ensure that no route information is lost in
  the conversion from list to trie.

* Fix a bug where an match on a route resource where the trie value is
  a RouteMatchOrVar would result in infinite recursion.

* Fix a bug where RVar trie value matching incorrectly matches on
  request paths that share a prefix with a valid trie key, but where
  the request paths do not contain a '/' following the matched prefix.

lseppala commented on src/Airship/Internal/Route.hs:
```diff
@@ -32,8 +40,37 @@ data BoundOrUnbound = Bound Text
+        -- in the desired manner. In the case of duplicate routes the
+        -- routes specified first are favored over any subsequent
+        -- specifications.
+        toTrie = foldr (uncurry insertOrReplace) Trie.empty
```
Since you're evaluating the entire list of routes, you'll want to use `foldl'`, which is more space and time efficient.

kellymclaughlin replied:
Can you explain this a bit more? Not saying your wrong, but I don't understand why there is a difference since `foldl'` is defined in terms of  `foldr`. Happy to change it, but I just copied this directly the `fromList` implementation in `Data.Trie`.

lseppala commented on src/Airship/Internal/Route.hs:
```diff
@@ -48,7 +48,29 @@ data RouteLeaf m = RouteMatch (Resource m) [Text]
+        -- specifications.
+        toTrie = foldr (uncurry insertOrReplace) Trie.empty
+        insertOrReplace k v t =
+            case Trie.member k t of
```
You could save a query by using `Trie.lookup` here instead:

```
Trie.insert k $ maybe v (`mergeValues` v) (Trie.lookup k t)
```
(alternatively, expand the `maybe` into the case statement if you'd prefer)

kellymclaughlin replied:
good idea

lseppala commented on src/Airship/Internal/Route.hs:
```diff
@@ -65,6 +102,42 @@ var t = Route [Var t]
+
+-- Routing trie creation algorithm
+-- 1. Store full paths as keys up to first `var`
+-- 2. Take SHA256 digest of the URL portion preceding the
```
Why are do we calculate the hash with SHA256? SHA256 is designed to be secure, and by nature of that property, it is not particularly fast. (see http://thepasswordproject.com/oclhashcat_benchmarking. SHA256 is more than 7x slower than MD5, for example)

I think what we want is for the hash to be _unique_, for which there are a lot faster non-secure algorithms. MurmurHash is generally a good one, but would require another dependency. `hashable` might work, though I know you need the `ByteString` for the Trie. I'll take a look and give you a suggestion.

Otherwise, MD4 (`Crypto.Hash.MD4`) is pretty fast.

lseppala replied:
Oh, `Data.Binary.encode`. 

```
let hash = Data.Binary.encode . Data.Hashable.hash :: Hashable a => a -> ByteString
```

lseppala replied:
Since you're only using the hash for a delimiter that the part of the path is a variable, you could instead use delimiter character. Rather than trying to find a printable character that's unused, I'd recommend using something like `'\US' :: Char`, which is the non-printable, ASCII Unit Separator character. `BC8.singleton '\US'`

kellymclaughlin replied:
Yeah the sha256 is probably overkill. I picked it because afaik there's still no known collisions for it and we were already building the dependency. The goal is just to have a unique key into the trie that will not conflict with another request path. I think actually we don't need a hash. I think base64 encoding will work just fine and then the risk of collisions dealing with *lesser* hashes is eliminated. 

I don't completely follow the final comment about the delimiter character. Are you suggesting storing the full route as a key with a delimiter substituted in for the `var` occurrences or something else? That would still require another delimiter character to be able to discern `var` from `star`.

lseppala commented on src/Airship/Internal/Route.hs:
```diff
@@ -76,40 +149,70 @@ star = Route [RestUnbound]
+            -- matched prefix is a RVar so calculate the key for the next part
+            -- of the route and continue attempting to match.
+            let nextKey = B.concat [ hash $ B.concat [matched, "var"]
+                                   , BC8.dropWhile (/='/') $ BC8.dropWhile (=='/') rest
```
This introduces a bit of inconsistency. 

The request `/foo//bar` will not match the route `/foo/bar`. But `/foo//bar` will match the route `/foo/:var`.

It's a separate discussion of whether we should drop repeat or trailing`/`s, but that would break backwards compatibility.

kellymclaughlin replied:
Right, this is another special case I forgot to handle. Will fix.

kellymclaughlin committed:
Make some changes based on review feedback

kellymclaughlin commented:
Latest commit has all the recommended changes except the `foldr` to `foldl'` change which I'd like to understand more before making.

kellymclaughlin committed:
s/foldr/foldl'

kellymclaughlin commented:
We have a need to merge this now for some internal projects we're working on. I have tests underway for this, but they are not quite finished. We will get those in prior to cutting a new official release that includes this work.
